core/server/lib/image/manipulator.js

```diff
@@ -61,4 +61,27 @@ const process = (options = {}) => {
         });
 };
 
+const resizeImage = (originalBuffer, {width, height} = {}) => {
+    const sharp = require('sharp');
+    return sharp(originalBuffer)
+        .resize(width, height, {
+            // CASE: dont make the image bigger than it was
+            withoutEnlargement: true
+        })
+        // CASE: Automatically remove metadata and rotate based on the orientation.
+        .rotate()
+        .toBuffer()
+        .then((resizedBuffer) => {
+            return resizedBuffer.length < originalBuffer.length ? resizedBuffer : originalBuffer;
+        });
+};
+
 module.exports.process = process;
+module.exports.safeResizeImage = (buffer, options) => {
+    try {
+        require('sharp');
+        return resizeImage(buffer, options);
+    } catch (e) {
+        return Promise.resolve(buffer);
```

Comment:
FYI design hint: libraries should return an error, callers need to decide what to do if an error is thrown/returned. Callers can ignore errors, callers can return errors to the next chain. Whatever...

---

Furthermore: If sharp isn't installed or resizeImage throws an error, we would save the same buffer stream as the resized image? Is that wanted? 🤔

Reply:
The lib returns a promise, which will reject if an error occurs, so the consumer of this is totally capable of catching any errors.

resizeImage returns promise, so errors from there will be handled by the library consumer - if sharp doesn't exist then yeah - we will save the resizedImage - i think this is a reasonable fallback for the functionality. @ErisDS what do you think?

Reply:
> The lib returns a promise, which will reject if an error occurs, so the consumer of this is totally capable of catching any errors.

IMO a library should always return a specific/custom/handled error. Otherwise, multiple/different callers need to handle an error from a third party library or from node. Plus the callers would need to wrap the error in case they want to forward the error.

Reply:
Hmm yeah - I think a specific error is good, I'm not sure how specific we can be on top of sharps errors though.

I noticed in this file we use InternalServer errors but that doesn't seem correct as the library has nothing to do with servers it just deals with image buffers and the filesystem :shrimp:

Do you think we should create some image manipulation errors to use in the module?

Reply:
> I noticed in this file we use InternalServer errors but that doesn't seem correct as the library has nothing to do with servers it just deals with image buffers and the filesystem 🦐

There are many definitions of what an internat server error is. 

> Do you think we should create some image manipulation errors to use in the module?

In general: We did not create custom error types for custom situations in the past, because

- an error is mostly http driven
- the most common http codes are reflected in our error set
- we can granulary define an error using "message", "type", "statusCode"

---

There is no need to block this PR, but it's important that all of us have a common thinking of how libraries are designed, how error handling works etc :) 

---

> resizeImage returns promise, so errors from there will be handled by the library consumer - if sharp doesn't exist then yeah - we will save the resizedImage - i think this is a reasonable fallback for the functionality. 

This just needs answering 👍